lib/pluginmanager/commands/install_command.rb

```diff
@@ -0,0 +1,108 @@
+# encoding: utf-8
+require "pluginmanager/command"
+require "bootstrap/util/compress"
+require "fileutils"
+require "pluginmanager/sources/http"
+require "pluginmanager/sources/local"
+
+class LogStash::PluginManager::InstallCommand < LogStash::PluginManager::Command
+
+  # Relocate a set of plugins within a pack directory into the logstash cache location. This gems
+  # are going to be used later on during installation time.
+  # @param pack_dir [String] The location of an uncompressed packaged plugin
+  def copy_packaged_gems(pack_dir)
+    FileUtils.cp(Dir.glob(File.join(pack_dir, "**", "*.gem")), LogStash::Environment::CACHE_PATH)
+  end
+
+  # Check if the installtion should be made in local mode, it's based on either having the local
+  # flag provided or on having packaged plugins to be installed.
+  # @return [Boolean] True in case of performing a local installation
+  def local?
+    @local || @packs
+  end
+
+  # Find in the extracted packaged plugins the list of plugins to be installed
+  # @param [String] The packaged plugin location in the file system, should be uncompressed.
+  # @return [Array] The list of plugins to be installed
+  def select_plugins(pack_dir)
+    Dir.glob(File.join(pack_dir, "*.gem")).map do |path|
+      fields = File.basename(path, ".gem").split("-")
+      (fields.count > 1 ? fields[0...-1] : fields).join("-")
+    end
+  end
+
+  # Fetch a package from a URI, either from a web uri or from the local file system.
+  # @param [URI] The package url
+  # @param [String] A temporary directy used to store the fetched package.
+  # @return [String] The location of the fetched package.
+  def fetch_pack(source, temp_dir)
+    file, _ =  source.fetch(temp_dir)
+    return file
+  end
+
+  # Extract a package of plugins
+  # @param [String] The package location
+  # @return [String] The location of the extracted content
+  def extract_pack(file)
+    filename = File.basename(file, ".zip").split("-")[0]
+    output_dir = File.join(File.dirname(file), "out")
+    zip.extract(file, output_dir)
+    return File.join(output_dir, "logstash", filename)
+  end
+
+  def find_packs(args=[])
+    packs = []
+    args.clone.each do |arg|
+      next if (arg.start_with?("logstash-") || File.extname(arg) == ".gem")
```

Comment:
at this point we are just sorting strings here, so having predicates is not easy unless we transform in earlier steps this strings to more complex objects.